ffi/derive/src/convert.rs

```diff
@@ -3,426 +3,483 @@ use proc_macro_error::{abort, OptionExt};
+                }
             }
-        }
 
-        impl #generic_params iroha_ffi::TryFromReprC<#lifetime> for &#lifetime mut #name #ty_generics #where_clause {
-            type Source = *mut iroha_ffi::Opaque;
-            type Store = Option<#name>;
+            impl<#generic_params> iroha_ffi::repr_c::CType<'_> for #name #ty_generics #where_clause {
+                type ReprC = *mut Self;
+                type RustStore = ();
+                type FfiStore = ();
 
-            unsafe fn try_from_repr_c(
-                source: <Self as iroha_ffi::TryFromReprC<#lifetime>>::Source,
-                store: &#lifetime mut <Self as iroha_ffi::TryFromReprC<#lifetime>>::Store
-            ) -> iroha_ffi::Result<Self> {
-                if source.is_null() {
-                    return iroha_ffi::FfiReturn::ArgIsNull;
+                fn into_repr_c(self, _: &mut ()) -> Self::ReprC {
+                    Box::into_raw(Box::new(self))
                 }
+                unsafe fn try_from_repr_c(source: Self::ReprC, _: &mut ()) -> iroha_ffi::Result<Self> {
+                    if let Some(_) = source.as_mut() {
+                        return Ok(*Box::from_raw(source));
+                    }
 
-                *store = Some(#name{__opaque_ptr: source});
-                store.as_mut().unwrap()
+                    Err(iroha_ffi::FfiReturn::ArgIsNull)
+                }
             }
+
         }
+    }
 
-        #opaque_item_slice_try_from_repr_c_derive
-        #opaque_item_vec_try_from_repr_c_derive
+    pub fn is_opaque_wrapper(input: &DeriveInput) -> bool {
+        let opaque_attr = parse_quote! {#[opaque_wrapper]};
+        input.attrs.iter().any(|a| *a == opaque_attr)
     }
 }
 
-fn derive_try_from_repr_c_for_opaque_item(input: &DeriveInput) -> TokenStream2 {
-    let name = &input.ident;
-    let mut generics = input.generics.clone();
-    let lifetime: syn::Lifetime = syn::parse_quote!('__iroha_ffi_itm);
-    let lifetimes = &[lifetime.clone()];
-    let (generic_params, ty_generics, where_clause) =
-        add_bounds_and_split_generics(&mut generics, lifetimes, &[], &[]);
-
-    let opaque_item_slice_try_from_repr_c_derive =
-        derive_try_from_repr_c_for_opaque_item_slice(input);
-    let opaque_item_vec_try_from_repr_c_derive = derive_try_from_repr_c_for_opaque_item_vec(input);
+mod enum_ {
+    use proc_macro2::TokenStream as TokenStream2;
+    use proc_macro_error::abort;
+    use quote::quote;
+    use syn::{parse_quote, Ident};
+
+    pub fn derive_ffi_type_for_fieldless_enum(
+        enum_name: &Ident,
+        enum_: &syn::DataEnum,
+        repr: &[syn::NestedMeta],
+    ) -> TokenStream2 {
+        let (ffi_type, largest_discriminant, is_wasm_primitive) = enum_size(enum_name, repr);
+        let (discriminants, discriminant_decls) = gen_discriminants(enum_name, enum_, &ffi_type);
+
+        // TODO: most of the times niche trait can be implemented and in edge cases robust trait
```

Comment:
So only case when enum does't have niche is when it has number of variants equal to `ffi_type::MAX - ffi_type::MIN`, we could check this condition during macro expansion.

Then we could do [the following](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b1fcf019c2454cff92e41b63f9d659fc) to find niche value.

Reply:
This warrants a separate issue.

Reply:
opened #2794